src/Microsoft.AspNet.BasicMiddleware/OverrideHeaderMiddleware.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.BasicMiddleware
+{
+    public class HeaderOverrideMiddleware
+    {
+        private OverrideHeaderMiddlewareOptions _options;
+        private RequestDelegate _next;
+        private const string XForwardedFor = "X-Forwarded-For";
+        private const string XForwardedHost = "X-Forwarded-Host";
+        private const string XForwardedProto = "X-Forwarded-Proto";
+        private const string XHttpMethodOverride = "X-Http-Method-Override";
+
+        public HeaderOverrideMiddleware(RequestDelegate next, OverrideHeaderMiddlewareOptions overrideMiddlewareOptions)
+        {
+            if (overrideMiddlewareOptions == null || next == null)
+            {
+                throw new ArgumentNullException(nameof(overrideMiddlewareOptions));
+            }
+            _options = overrideMiddlewareOptions;
+            _next = next;
+        }
+
+        /// <summary>
+        ///  Read the values for the different header keys we are looking for
+        /// then overrode those values in the request with the values in the header
+        ///     - X-Forward-For
+        ///     - X-Forwarded-Host
+        ///     - X-Forwarded-Proto
+        ///     - X-Http-Method-Override
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            var xForward = context.Request.Headers.GetCommaSeparatedValues(XForwardedFor);
+            if (_options.XForwardedForEnabled && xForward != null && xForward.Count > 0)
+            {
+                IPAddress originalIPAddress;
+                if (IPAddress.TryParse(xForward[0], out originalIPAddress))
```

Comment:
Is the XFF header the only security concern?

What if the client tries to spoof the protocol/host/etc... using some of the other X-* headers? Or are we relying on the proxy to stomp the relevant X-* headers since this middleware will be opt-in.